// @sentry/wasm/esm/index.js
import { defineIntegration, convertIntegrationFnToClass } from '@sentry/core';
import { patchWebAssembly } from './patchWebAssembly.js';
import { getImages, getImage } from './registry.js';

const INTEGRATION_NAME = 'Wasm';

const _wasmIntegration = (() => {
  return {
    name: INTEGRATION_NAME,
    setupOnce() {
      patchWebAssembly();
    },
    processEvent(event) {
      let haveWasm = false;

      if (event.exception && event.exception.values) {
        event.exception.values.forEach(exception => {
          if (exception.stacktrace && exception.stacktrace.frames) {
            haveWasm = haveWasm || patchFrames(exception.stacktrace.frames);
          }
        });
      }

      if (haveWasm) {
        event.debug_meta = event.debug_meta || {};
        event.debug_meta.images = [...(event.debug_meta.images || []), ...getImages()];
      }

      return event;
    },
  };
}) ;

const wasmIntegration = defineIntegration(_wasmIntegration);

/**
 * Process WASM stack traces to support server-side symbolication.
 *
 * This also hooks the WebAssembly loading browser API so that module
 * registrations are intercepted.
 *
 * @deprecated Use `wasmIntegration` export instead
 *
 * import { wasmIntegration } from '@sentry/wasm';
 *
 * ```
 * Sentry.init({ integrations: [wasmIntegration()] });
 * ```
 */
// eslint-disable-next-line deprecation/deprecation
const Wasm = convertIntegrationFnToClass(INTEGRATION_NAME, wasmIntegration)

;

/**
 * Patches a list of stackframes with wasm data needed for server-side symbolication
 * if applicable. Returns true if any frames were patched.
 */
function patchFrames(frames) {
  let haveWasm = false;
  frames.forEach(frame => {
    if (!frame.filename) {
      return;
    }
    const match = frame.filename.match(/^(.*?):wasm-function\[\d+\]:(0x[a-fA-F0-9]+)$/);
    if (match !== null) {
      const index = getImage(match[1]);
      if (index >= 0) {
        frame.instruction_addr = match[2];
        frame.addr_mode = `rel:${index}`;
        frame.filename = match[1];
        frame.platform = 'native';
        haveWasm = true;
      }
    }
  });
  return haveWasm;
}

export { Wasm, wasmIntegration };
