// @bdelab/roar-firekit/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initializeFirebaseProject = exports.getTreeTableOrgs = exports.emptyOrgList = exports.emptyOrg = exports.RoarTaskVariant = exports.RoarAppUser = exports.RoarAppkit = exports.RoarFirekit = void 0;
var firekit_1 = require("./firestore/firekit");
Object.defineProperty(exports, "RoarFirekit", { enumerable: true, get: function () { return firekit_1.RoarFirekit; } });
var appkit_1 = require("./firestore/app/appkit");
Object.defineProperty(exports, "RoarAppkit", { enumerable: true, get: function () { return appkit_1.RoarAppkit; } });
var user_1 = require("./firestore/app/user");
Object.defineProperty(exports, "RoarAppUser", { enumerable: true, get: function () { return user_1.RoarAppUser; } });
var task_1 = require("./firestore/app/task");
Object.defineProperty(exports, "RoarTaskVariant", { enumerable: true, get: function () { return task_1.RoarTaskVariant; } });
var util_1 = require("./firestore/util");
Object.defineProperty(exports, "emptyOrg", { enumerable: true, get: function () { return util_1.emptyOrg; } });
Object.defineProperty(exports, "emptyOrgList", { enumerable: true, get: function () { return util_1.emptyOrgList; } });
Object.defineProperty(exports, "getTreeTableOrgs", { enumerable: true, get: function () { return util_1.getTreeTableOrgs; } });
Object.defineProperty(exports, "initializeFirebaseProject", { enumerable: true, get: function () { return util_1.initializeFirebaseProject; } });
__exportStar(require("./firestore/query-assessment"), exports);
